demos/selenium-test-app/lib.rs

```diff
@@ -0,0 +1,239 @@
+use crate::AlternativeOriginsMode::{CertifiedContent, Redirect};
+use candid::{CandidType, Deserialize, Principal};
+use ic_cdk::api;
+use ic_cdk_macros::{init, post_upgrade, query, update};
+use ic_certified_map::{labeled_hash, AsHashTree, Hash, RbTree};
+use lazy_static::lazy_static;
+use serde::Serialize;
+use serde_bytes::{ByteBuf, Bytes};
+use sha2::Digest;
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+thread_local! {
+    static ASSETS: RefCell<HashMap<&'static str, (Vec<(String, String)>, Vec<u8>)>> = RefCell::new(HashMap::default());
+    static ASSET_HASHES: RefCell<RbTree<&'static str, [u8; 32]>> = RefCell::new(RbTree::default());
+    static ALTERNATIVE_ORIGINS_MODE: RefCell<AlternativeOriginsMode> = RefCell::new(CertifiedContent);
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ContentType {
+    HTML,
+    JS,
+    JSON,
+}
+
+impl ContentType {
+    pub fn to_mime_type_string(&self) -> String {
+        match self {
+            ContentType::HTML => "text/html".to_string(),
+            ContentType::JS => "text/javascript".to_string(),
+            ContentType::JSON => "application/json".to_string(),
+        }
+    }
+}
+
+#[query]
+fn whoami() -> Principal {
+    api::caller()
+}
+
+#[update]
+fn update_alternative_origins(content: String, mode: AlternativeOriginsMode) {
+    ASSETS.with(|a| {
+        ASSET_HASHES.with(|ah| {
+            let mut assets = a.borrow_mut();
+            let mut asset_hashes = ah.borrow_mut();
+            let path = "/.well-known/ii-alternative-origins";
+            assets.insert(
+                path,
+                (
+                    vec![(
+                        "Content-Type".to_string(),
+                        ContentType::JSON.to_mime_type_string(),
+                    )],
+                    content.as_bytes().to_vec(),
+                ),
+            );
+            asset_hashes.insert(path, sha2::Sha256::digest(content.as_bytes()).into());
+            update_root_hash(&asset_hashes);
+        });
+    });
+
+    ALTERNATIVE_ORIGINS_MODE.with(|m| {
+        m.replace(mode);
+    })
+}
+
+pub type HeaderField = (String, String);
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub struct HttpRequest {
+    pub method: String,
+    pub url: String,
+    pub headers: Vec<(String, String)>,
+    pub body: ByteBuf,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub struct HttpResponse {
+    pub status_code: u16,
+    pub headers: Vec<HeaderField>,
+    pub body: Cow<'static, Bytes>,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub enum AlternativeOriginsMode {
+    CertifiedContent,
+    UncertifiedContent,
+    Redirect { location: String },
+}
+
+#[query]
+pub fn http_request(req: HttpRequest) -> HttpResponse {
```

Comment:
isn't this better implemented as an asset canister? why re-implement the asset serving logic?

Reply:
How exactly do we tamper with the behavior? Also that's a big-ish design decision (for this test app), can we record that somewhere in the README?

Reply:
Yes, see #751.